tests/testsuite/rustflags.rs

```diff
@@ -993,6 +1054,188 @@ fn target_rustflags_normal_source() {
         .run();
 }
 
+#[cargo_test]
+fn target_rustflags_also_for_build_scripts() {
+    let p = project()
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+                fn main() { }
+                #[cfg(not(foo))]
+                fn main() { }
+            "#,
+        )
+        .file(
+            ".cargo/config",
+            &format!(
+                "
+            [target.{}]
+            rustflags = [\"--cfg=foo\"]
+            ",
+                rustc_host()
+            ),
+        )
+        .build();
+
+    p.cargo("build").run();
+}
+
+#[cargo_test]
+fn target_rustflags_not_for_build_scripts_with_target() {
+    let host = rustc_host();
+    let p = project()
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+                fn main() { }
+                #[cfg(foo)]
+                fn main() { }
+            "#,
+        )
+        .file(
+            ".cargo/config",
+            &format!(
+                "
+            [target.{}]
+            rustflags = [\"--cfg=foo\"]
+            ",
+                host
+            ),
+        )
+        .build();
+
+    p.cargo("build --target").arg(host).run();
+
+    // Enabling -Ztarget-applies-to-host should not make a difference without the config setting
+    p.cargo("build --target")
+        .arg(host)
+        .masquerade_as_nightly_cargo()
+        .arg("-Ztarget-applies-to-host")
+        .run();
+
+    // Even with the setting, the rustflags from `target.` should not apply, to match the legacy
+    // Cargo behavior.
+    p.change_file(
+        ".cargo/config",
+        &format!(
+            "
+        target-applies-to-host = true
+
+        [target.{}]
+        rustflags = [\"--cfg=foo\"]
+        ",
+            host
+        ),
+    );
+    p.cargo("build --target")
+        .arg(host)
+        .masquerade_as_nightly_cargo()
+        .arg("-Ztarget-applies-to-host")
+        .run();
```

Comment:
I mentioned this above, but this doesn't match my own mental model at least because if `target-applies-to-host = true` then I'd expect that target-specific flags apply to the host builds.

Reply:
Ah, so, this one is special, and gets back to my point about how `target-applies-to-host` should _arguably_ be a tri-state of `true/false/legacy`. _Currently_ (as in, on `master`), if you pass `--target <host triple>`, then build scripts will _not_ pick up `target.<host>.rustflags` due to

https://github.com/rust-lang/cargo/blob/c7dbc422bbf036b5c48e1accaa1e74cac0870544/src/cargo/core/compiler/build_context/target_info.rs#L614-L619

Example (using latest Cargo release):

```console
$ cargo new bs
     Created binary (application) `bs` package
$ cd bs
$ echo 'fn main() { assert!(cfg!(foo)); }' > build.rs
$ mkdir .cargo
$ echo '[target.x86_64-apple-darwin]' > .cargo/config
$ echo 'rustflags = ["--cfg=foo"]' >> .cargo/config
$ cargo b
   Compiling bs v0.1.0 (/Users/jongje/bs)
    Finished dev [unoptimized + debuginfo] target(s) in 1.27s
$ cargo b --target x86_64-apple-darwin
   Compiling bs v0.1.0 (/Users/jongje/bs)
error: failed to run custom build command for `bs v0.1.0 (/Users/jongje/bs)`

Caused by:
  process didn't exit successfully: `/Users/jongje/bs/target/debug/build/bs-7d41676a7672d1a0/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at 'assertion failed: cfg!(foo)', build.rs:1:13
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

if `target-applies-to-host` is just a `bool`, and `true` is supposed to be the _existing_ behavior, then making `target.<host>.rustflags` apply to build scripts with `--target <host>` would be a backwards-incompatible change (as per above). Now, maybe that's okay, and we want to consider that fixing a bug — that's up to you. Otherwise, we'd need a third state for `target-applies-to-host` that maps exactly to the current behavior.